server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,440 @@
+    private final AtomicBoolean shouldThrottleIncomingMerges = new AtomicBoolean();
+    // how many {@link MergeTask}s have kicked off (this is used to name them).
+    private final AtomicLong mergeTaskCount = new AtomicLong();
+    private final AtomicLong mergeTaskDoneCount = new AtomicLong();
+
+    public ThreadPoolMergeScheduler(ShardId shardId, IndexSettings indexSettings, ThreadPoolMergeQueue threadPoolMergeQueue) {
+        this.shardId = shardId;
+        this.config = indexSettings.getMergeSchedulerConfig();
+        this.logger = Loggers.getLogger(getClass(), shardId);
+        this.mergeTracking = new MergeTracking(
+            logger,
+            () -> this.config.isAutoThrottle() ? threadPoolMergeQueue.getTargetMBPerSec() : Double.POSITIVE_INFINITY
+        );
+        this.threadPoolMergeQueue = threadPoolMergeQueue;
+    }
+
+    @Override
+    public Set<OnGoingMerge> onGoingMerges() {
+        return mergeTracking.onGoingMerges();
+    }
+
+    @Override
+    public MergeStats stats() {
+        return mergeTracking.stats();
+    }
+
+    @Override
+    public MergeScheduler getMergeScheduler() {
+        return this;
+    }
+
+    @Override
+    public void refreshConfig() {
+        // if maxMergeCount changed, maybe we need to toggle merge task throttling
+        checkMergeTaskThrottling();
+        // if maxThreadCount changed, maybe some backlogged merges are now allowed to run
+        enqueueBacklogged();
+    }
+
+    @Override
+    public void merge(MergeSource mergeSource, MergeTrigger trigger) throws IOException {
+        MergePolicy.OneMerge merge = mergeSource.getNextMerge();
+        if (merge != null) {
+            submitNewMergeTask(mergeSource, merge, trigger);
+        }
+    }
+
+    /**
+     * A callback allowing for custom logic before an actual merge starts.
+     */
+    protected void beforeMerge(OnGoingMerge merge) {}
+
+    /**
+     * A callback allowing for custom logic after an actual merge starts.
+     */
+    protected void afterMerge(OnGoingMerge merge) {}
+
+    protected void enableMergeTaskThrottling(int numRunningMerges, int numQueuedMerges, int configuredMaxMergeCount) {}
+
+    protected void disableMergeTaskThrottling(int numRunningMerges, int numQueuedMerges, int configuredMaxMergeCount) {}
+
+    @Override
+    public MergeScheduler clone() {
+        // Lucene IW makes a clone internally but since we hold on to this instance
+        // the clone will just be the identity.
+        return this;
+    }
+
+    protected void handleMergeException(Throwable t) {
+        throw new MergePolicy.MergeException(t);
+    }
+
+    private void submitNewMergeTask(MergeSource mergeSource, MergePolicy.OneMerge merge, MergeTrigger mergeTrigger) {
+        MergeTask mergeTask = newMergeTask(mergeSource, merge, mergeTrigger);
+        assert mergeTask.isRunning() == false;
+        threadPoolMergeQueue.enqueueMergeTask(mergeTask);
+        threadPoolMergeQueue.executeNextMergeTask();
```

Comment:
I think I would prefer just one call to the merge queue, perhaps "submitMergeTask".

Reply:
Pushed [c5b05e5](https://github.com/elastic/elasticsearch/pull/120869/commits/c5b05e56afa2bfcce6dbac26074b8fa75581260f)